File: node/src/runtime/os-profile.ts
import type { GovernancePolicy } from "../governance.js"

export type OsProfileId = "native"

export interface NativeOsProfile {
  id: OsProfileId
  attentionPolicy: { maxQueueSize?: number }
  governancePolicy: GovernancePolicy
}

/** Default attention policy for native profile smoke tests. */
export const DEFAULT_NATIVE_ATTENTION_POLICY = { maxQueueSize: 64 }

/** Permissive governance policy for native runs that do not need AskUser. */
export const DEFAULT_NATIVE_GOVERNANCE_POLICY: GovernancePolicy = {
  rules: [{ pattern: "*", action: "allow" }],
}

/** Default restrictive sandbox policy template requiring confirmation for modification/execution. */
export const DEFAULT_SANDBOX_POLICY: GovernancePolicy = {
  rules: [
    { pattern: "read_file", action: "allow" },
    { pattern: "write_file", action: "ask_user" },
    { pattern: "run_command", action: "ask_user" },
    { pattern: "*", action: "deny" },
  ],
}

/** Resolve a named OS profile into concrete kernel-owned policy defaults. */
export function osProfile(profile: OsProfileId | NativeOsProfile = "native"): NativeOsProfile {
  if (typeof profile !== "string") return profile
  if (profile !== "native") throw new Error(`Unsupported OS profile: ${profile}`)
  return {
    id: "native",
    attentionPolicy: DEFAULT_NATIVE_ATTENTION_POLICY,
    governancePolicy: DEFAULT_NATIVE_GOVERNANCE_POLICY,
  }
}

/** Assert that a runtime is using a valid native microkernel policy profile. */
export function assertNativeProfile(profile: OsProfileId | NativeOsProfile = "native"): NativeOsProfile {
  const resolved = osProfile(profile)
  if (resolved.id !== "native") {
    throw new Error(`Unsupported OS profile: ${resolved.id}`)
  }
  const validation = validateDeclarativePolicy(resolved.governancePolicy, resolved.attentionPolicy)
  if (!validation.valid) {
    throw new Error(`Invalid native OS profile: ${validation.errors.join("; ")}`)
  }
  return resolved
}

/**
 * Validates the declarative policies statically to prevent runtime crashes when loaded into the microkernel.
 */
export function validateDeclarativePolicy(
  govPolicy?: GovernancePolicy,
  attentionPolicy?: { maxQueueSize?: number },
): { valid: boolean; errors: string[] } {
  const errors: string[] = []

  if (govPolicy) {
    if (!Array.isArray(govPolicy.rules)) {
      errors.push("GovernancePolicy rules must be an array")
    } else {
      govPolicy.rules.forEach((rule, idx) => {
        if (!rule.pattern || typeof rule.pattern !== "string") {
          errors.push(`Rule[${idx}] pattern is missing or not a string`)
        }
        if (!["allow", "deny", "ask_user"].includes(rule.action)) {
          errors.push(`Rule[${idx}] action '${rule.action}' is invalid. Allowed: allow, deny, ask_user`)
        }
      })
    }
  }

  if (attentionPolicy) {
    if (attentionPolicy.maxQueueSize !== undefined) {
      if (typeof attentionPolicy.maxQueueSize !== "number" || attentionPolicy.maxQueueSize <= 0) {
        errors.push("AttentionPolicy maxQueueSize must be a positive integer")
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  }
}

File: rust/src/runtime/os_profile.rs
pub use deepstrike_core::runtime::kernel::wire::{
    ParamConstraint as ConstraintSpec, PolicyAction, PolicyRule, RateLimitSpec,
};
use deepstrike_core::runtime::kernel::wire::{SignalPolicy as WireSignalPolicy, WireU64};
pub use deepstrike_core::scheduler::policy::SchedulerPolicyConfig;

use crate::{Error, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SignalPolicy {
    pub queue_max: u32,
    pub ttl_ms: Option<u64>,
    pub deadline_escalation: Option<bool>,
}

impl SignalPolicy {
    pub(crate) fn into_kernel(self) -> WireSignalPolicy {
        WireSignalPolicy {
            queue_max: self.queue_max,
            ttl_ms: self.ttl_ms.map(WireU64::new),
            deadline_escalation: self.deadline_escalation,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MemoryWriteRateLimit {
    pub max_writes: u32,
    pub window_ms: u64,
}

impl From<MemoryWriteRateLimit> for (u32, u64) {
    fn from(limit: MemoryWriteRateLimit) -> Self {
        (limit.max_writes, limit.window_ms)
    }
}

#[derive(Debug, Clone)]
pub struct GovernancePolicy {
    pub default_action: Option<PolicyAction>,
    pub rules: Vec<PolicyRule>,
    pub vetoed_tools: Vec<String>,
    pub rate_limits: Vec<RateLimitSpec>,
    pub constraints: Vec<ConstraintSpec>,
    /// I5: when true (default), the kernel withholds statically denied tools (vetoes and `deny`
    /// rules, evaluated exactly as the call gate evaluates them) from the provider surface and
    /// names them once in the knowledge slot. Mirrors Node.
    pub surface_denied_in_system: bool,
}

impl GovernancePolicy {
    pub fn allow_all() -> Self {
        Self {
            default_action: None,
            rules: vec![PolicyRule {
                tool_pattern: "*".to_string(),
                action: PolicyAction::Allow,
            }],
            vetoed_tools: vec![],
            rate_limits: vec![],
            constraints: vec![],
            surface_denied_in_system: true,
        }
    }

    pub(crate) fn into_host_fact(self) -> serde_json::Value {
        serde_json::json!({
            "kind": "load_governance_policy",
            "default_action": self.default_action,
            "rules": self.rules,
            "vetoed_tools": self.vetoed_tools,
            "rate_limits": self.rate_limits,
            "constraints": self.constraints,
            "hide_denied_tools": self.surface_denied_in_system,
        })
    }

    /// A §13.2 live-policy patch replacing the governance posture, for
    /// [`crate::RuntimeRunner::apply_policy_patch`].
    pub fn into_policy_patch(self) -> serde_json::Value {
        let mut policy = self.into_host_fact();
        let object = policy.as_object_mut().expect("the host fact is an object");
        object.remove("kind");
        if object
            .get("default_action")
            .is_some_and(serde_json::Value::is_null)
        {
            object.remove("default_action");
        }
        serde_json::json!({ "kind": "replace_governance_policy", "policy": policy })
    }
}

#[derive(Debug, Clone)]
pub struct NativeOsProfile {
    pub id: &'static str,
    pub signal_policy: SignalPolicy,
    pub governance_policy: GovernancePolicy,
}

#[derive(Debug, Clone)]
pub enum OsProfile {
    Native,
    Concrete(NativeOsProfile),
}

pub const DEFAULT_NATIVE_SIGNAL_POLICY: SignalPolicy = SignalPolicy {
    queue_max: 64,
    ttl_ms: None,
    deadline_escalation: None,
};

pub fn default_native_governance_policy() -> GovernancePolicy {
    GovernancePolicy::allow_all()
}

pub fn os_profile(profile: Option<OsProfile>) -> NativeOsProfile {
    match profile.unwrap_or(OsProfile::Native) {
        OsProfile::Native => NativeOsProfile {
            id: "native",
            signal_policy: DEFAULT_NATIVE_SIGNAL_POLICY,
            governance_policy: default_native_governance_policy(),
        },
        OsProfile::Concrete(profile) => profile,
    }
}

pub fn assert_native_profile(profile: Option<OsProfile>) -> Result<NativeOsProfile> {
    let resolved = os_profile(profile);
    if resolved.id != "native" {
        return Err(Error::Other(format!(
            "Unsupported OS profile: {}",
            resolved.id
        )));
    }
    if resolved.signal_policy.queue_max == 0 {
        return Err(Error::Other(
            "Invalid native OS profile: SignalPolicy queue_max must be positive".to_string(),
        ));
    }
    if matches!(resolved.signal_policy.ttl_ms, Some(0)) {
        return Err(Error::Other(
            "Invalid native OS profile: SignalPolicy ttl_ms must be positive when present"
                .to_string(),
        ));
    }
    Ok(resolved)
}
